Add new query args to Pattern Directory Controller

ntsekouras opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/57501


Adds the 'per_page', 'page', 'offset', 'order', and 'orderby' args to WP_REST_Pattern_Directory_Controller.

GB PR: https://github.com/WordPress/gutenberg/pull/45293

--cc @hellofromtonya @ryelle 

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ntsekouras committed:
Add new query args to Pattern Directory Controller

hellofromtonya commented on src/wp-includes/rest-api/endpoints/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -93,31 +94,19 @@ public function get_items( $request ) {
-		$query_args = array(
-			'locale'     => get_user_locale(),
-			'wp-version' => $wp_version,
+		$valid_query_args = array( 'offset', 'order', 'orderby', 'page', 'per_page', 'search', 'slug' );
+		$query_args       = array_merge(
+			array_intersect_key( $request->get_params(), array_flip( $valid_query_args ) ),
+			array(
+				'locale'     => get_user_locale(),
+				'wp-version' => $wp_version, // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UndefinedVariable -- it's defined in `version.php` above.
+			)
 		);
```
This code can be simplified and made more performant:

```suggestion
		$valid_query_args = array(
			'offset'   => true,
			'order'    => true,
			'orderby'  => true,
			'page'     => true,
			'per_page' => true,
			'search'   => true,
			'slug'     => true,
		);
		$query_args               = array_intersect_key( $request->get_params(), $valid_query_args );
		$query_args['locale']     = get_user_locale();
		$query_args['wp-version'] = $wp_version;
```

The `array_flip()` creates an array with the values of `$valid_query_args`  as the keys. Since this is used only for filtering the args, why not set up the `$valid_query_args` into a keyed structure? In doing so, the `array_flip()` is eliminated.

The `array_merge()` also creates another array. As `'locale'` and `'wp-version'` keys will not be in the array after doing the `array_intersect_key()`, a more performant way is to directly adding them to `$query_args` rather than `array_merge()`.

The changes remove 2 unnecessary array iterations and creations.

hellofromtonya replied:
Fixed in https://github.com/WordPress/wordpress-develop/pull/3861/commits/1cede3c669fdfb0588f2c5562db0fa6c54c10aae.

hellofromtonya replied:
See it in action https://3v4l.org/DGvja

ntsekouras replied:
Thanks @hellofromtonya !

hellofromtonya committed:
Simplify and tweak performance of  query_args construction

hellofromtonya committed:
Removes gutenberg text context

hellofromtonya committed:
Add params as keys to test dataset

hellofromtonya committed:
Adds assert messages to identify which assertion failed

hellofromtonya committed:
Make clear that "pre_http_request" is a mocked callback

Reuses pattern from other "pre_http_request" mocks to
convert the static function into a public callback.

hellofromtonya reviewed (approved):
The code itself LGTM 👍 

Next step is to test it and add a Test Report. If all is good, then commit.

hellofromtonya commented:
Once committed, I can open a PR in Gutenberg to synch the changes from here back to Gutenberg.

mukeshpanchal27 commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -310,6 +340,174 @@ static function( $response ) {
+	/**
+	 * Tests if the provided query args are passed through to the wp.org API.
+	 *
+	 * @dataProvider data_get_items_query_args
+	 *
+	 * @covers WP_REST_Pattern_Directory_Controller::get_items
+	 *
+	 * @since 6.2.0
```
Per [56070](https://core.trac.wordpress.org/ticket/56070) proposal, the document formate needs to changes for consistency. 

```suggestion
	 * @since 6.2.0
	 *
	 * @ticket 57501
	 *
	 * @covers WP_REST_Pattern_Directory_Controller::get_items
	 *
	 * @dataProvider data_get_items_query_args
```

hellofromtonya replied:
Noting: that proposal has not been adopted yet as it's still in discussion. That said, it's okay to go ahead with the changes here should the proposal be adopted.

mukeshpanchal27 commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -310,6 +340,174 @@ static function( $response ) {
+	}
+
+	/**
+	 * Data provider.
```
```suggestion
	 * Data provider for test_get_items_query_args().
```

hellofromtonya replied:
I think this is unnecessary. The naming convention of `test_` and `data_` align to know that this data provider is for the matching test.

hellofromtonya replied:
Whoops forgot to share where the naming convention is identified the handbook 🤦‍♀️ 

>A [data provider](https://phpunit.readthedocs.io/en/9.5/writing-tests-for-phpunit.html#data-providers) is a secondary function which is linked to a test using a` @dataProvider `annotation in the docblock of the test method and is generally named after the test, replacing the `test` prefix in the method name with `data`.

https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#repetitive-tests.

mukeshpanchal27 commented on tests/phpunit/tests/rest-api/rest-pattern-directory-controller.php:
```diff
@@ -573,4 +771,33 @@ static function ( $response, $parsed_args, $url ) use ( $blocked_host ) {
 		);
 	}
+
+	/**
+	 * Mock the request to wp.org URL to capture the URLs.
+	 *
+	 * @since 6.2.0
+	 *
+	 * @return array faux/mocked response.
+	 */
+	public function mock_request_to_apiwporg_url( $response, $args, $url ) {
```
In the core, some mock functions have the document and some don't. I appreciated @hellofromtonya's and @SergeyBiryukov's input on this. 

hellofromtonya replied:
I know. The handbook and similar test mocks needs some love to align for consistency. I agree!

mukeshpanchal27 reviewed (commented):
Thanks @ntsekouras, Left some feedback on unit test.

hellofromtonya committed:
Rearrangement of test annotations

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/55132.
